Add new WebSocket interface to Broker (#219)
zeek/broker: doc/web-socket.rst

.. _web-socket:

Web Socket
==========

Broker offers access to the publish/subscribe layer via WebSocket in order to
make its :ref:`data <data-model>` accessible to third parties.

WebSocket clients are treated as lightweight peers. Each Broker endpoint can be
configured to act as a WebSocket server by either setting the environment
variable ``BROKER_WEB_SOCKET_PORT`` or by setting ``broker.web-socket.port`` on
the command line or in the configuration file. When running Zeek, scripts may

> also redefine ``Broker::web_socket_port`` to a valid TCP port.

Reviewer: ```suggestion
also redefine ``Broker::default_port_websocket`` to a valid TCP port.
```

That's how my Zeek-side branch now names this.

Author: I've changed this in [5b26424](https://github.com/zeek/broker/pull/219/commits/5b26424a823fb12cda46e9e009bb97e6ce97c793).

endpoint, Broker allows JSON-formatted text messages only.

Handshake
~~~~~~~~~

The broker endpoint expects a JSON array of strings as the first message. This
array encodes the subscriptions as a list of topic prefixes that the client
subscribes to. Clients that only publish data must send an empty JSON array.

After receiving the subscriptions, the Broker endpoint sends a single ACK
message:

.. code-block:: json

  {
    "type": "ack",
    "endpoint": "<uuid>",
    "version": "<broker-version>"
  }

In this message, ``<uuid>`` is the unique endpoint ID of the WebSocket server
and ``<broker-version>`` is a string representation of the ``libbroker``
version, i.e., the result of ``broker::version::string()``. For example:

.. code-block:: json

  {
    "type": "ack",
    "endpoint": "925c9110-5b87-57d9-9d80-b65568e87a44",
    "version": "2.2.0-22"
  }

Protocol
~~~~~~~~

After the handshake, the WebSocket client may only send `Data Messages`_. The
Broker endpoint converts every message to its native representation and
publishes it.

The WebSocket server may send `Event Messages`_ (whenever a data message matches

> the subscriptions of the client) and `Error Messages_` to the client.

Reviewer: I think the distinction between data messages and event message isn't clear here. If we are in pure Broker-land (no Zeek), aren't then both sides only sending generic data messages? I.e., no event semantics. If we are talking Zeek, then both sides are sending events, which have a specific format. I would suggest to make that distinction explicit: talk about the two cases and (1) show generic messages in the first case, and (2) show concrete examples of how a Zeek event looks liken the 2nd. The latter will then also help people writing Zeek clients to know the event format, because that's currently not documented anywhere I believe (on the C++ side, we got a wrapper hiding that iirc).

Additional thought: maybe using `event` as the message type isn't even quite right at the Broker layer? And more generally, can we make the two sides symmetrical so that client and server send their data in the same format (data messages I guess)?

Reviewer: Looks good, I adapted the agent and added a Zeek-side test in https://github.com/zeek/zeek/pull/2121. All working well.

Data Representation
~~~~~~~~~~~~~~~~~~~

Broker uses a recursive data type to represent its values (see
:ref:`Data Model <data-model>`). This data model does not map to JSON-native
types without ambiguity, e.g., because Broker distinguishes between signed and
unsigned number types.

In JSON, we represent each value as a JSON object with two keys: ``@data-type``
and ``data``. The former identifies one of Broker's data types (see below) and
denotes how Broker parses the ``data`` field.

None
****

There is only exactly one valid input for encoding a ``none``:

.. code-block:: json

  {
    "@data-type": "none",
    "data": {}
  }

Boolean
*******

The type ``boolean`` can take on exactly two values and maps to the native JSON
boolean type:

.. code-block:: json

  {

>     "@data-type": "boolean",

Reviewer: Why's this prefixed with `@`, whereas other keys aren't? (Just curious).

Author: The idea is to flag meta fields with an `@` since they are more like annotations than actual fields. In this case, this is an annotation for the `data` field that tells the parser what type to use.

    "code": "deserialization_failed",
    "context": "input #1 contained malformed JSON -> caf::pec::unexpected_character(1, 1)"
  }

Encoding of Zeek Events
~~~~~~~~~~~~~~~~~~~~~~~

Broker encodes Zeek events as nested vectors using the following structure:
``[<format-nr>, <type>, [<name>, <args>]]``:

``format-nr``
  A ``count`` denoting the format version. Currently, this is always ``1``.

``type``
  A ``count`` denoting the encoded Zeek message type. For events, this is always
  ``1``. Other message types in Zeek are currently not safe for 3rd-party use.

``name``
  Identifies the Zeek event.

``args``

>   Contains the arguments for the event. Usually, this is another ``vector``.

Reviewer: ```suggestion
  Contains the arguments for the event in the form of another ``vector``.
```
